### PR TITLE
SJSU-1011, Shopping Cart

### PR DESCRIPTION
**Pull request by ynoe**
Story: Create a Shopping Cart
What I did:
Implemented count tracking for products in cart and an array of distinct products 
implemented increment and decrement product count in shopping cart
implemented delete all for a product(with any count >= 1) in cart
Implemented wiring to checkout via checkout button

*Installation needed on your local machine, cd to front_end under sjsu_shop:
$ npm install --save react react-dom 
$ npm install --save react-bootstrap

To Verify:
Suggestions for steps for testing:
1.	Click on shopping cart icon on the navbar: Shopping Cart is empty
2.	Click on Categories or search, click on add to cart buttons OR click on item name which directs to product detail page, and click on add to cart button
3.	Click on cart icon on navbar: Shopping cart should have correct count for each product
4.	Delete in any sequence, from front/middle/end of array, with any delete/decrement button and nothing breaks. Increment should still work.
5.	Click on checkout button, redirects user to checkout page and displays correct messages/product details. 
6.	Shopping cart modal should disappear when on checkout page
7.	Click on shopping cart icon, shopping cart modal should appear for user to easily edit products in shopping cart while on the checkout page.
8.	User makes edits on shopping cart modal. User clicks checkout, redirects user to checkout page and shopping cart modal should disappear again.
9.	If user empties cart, cart should become empty. 
10.	Repeat any combination of above steps and nothing should break.

How to get array of cart with products for other features etc. storing for user profile:
Replace YourOwnClassName, handle your own conditions (since this implementation allows for best code reuse, but will send data to connected js pages whenever reducer state changes) 

import { connect } from "react-redux";

render() {
    const cart = this.props.cart.shoppingCart;

const mapStateToProps = state => ({
  cart: state.cart
});

//connect to cartReducer to display items in cart
export default connect(
  mapStateToProps,
  {}
)(YourOwnClassName);

### CODE REVIEW

File: front_end/src/components/cart/ShoppingCartModal.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    super();
+    this.state = {
+      redirect: false
+    };
+  }
+
+  render() {
+    if (this.state.redirect) {
+      this.state.redirect = false;
+      return <Redirect push to="/checkout" />;
```
Comment (minhphan156) [inline front_end/src/components/cart/ShoppingCartModal.js]: This is cool. I dont know we have this API in react-router.

File: front_end/src/reducers/cartReducer.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      //If cart array empty OR there is NO occurrence of this product being
+      //processed, then this code gets executed
+      action.payload[0]["count"] = "1"; //Product object gets field count set to 1
+      return {
+        ...state,
+        shoppingCart: [...state.shoppingCart, action.payload[0]]
+      };
+
+    case REMOVE_ITEM:
+      if (action.payload.productId == "All Product Ids") {
```
[the remark this answers is not in the document]
Comment (ynoe) [reply front_end/src/reducers/cartReducer.js]: Resolved in CartReducer.js and ShoppingCart.js

File: front_end/src/components/cart/ShoppingCart.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            <span className="cart-modal-subTotalBox font-weight-bold">
+              Subtotal:
+            </span>
+            <span> ${(total / 100).toFixed(2)}</span>
+          </div>
+          <div className="btn-group d-flex justify-content-center" role="group">
+            <button
+              className="btn btn-light"
+              onClick={() =>
+                this.onDecrementCountClick("All Product Ids", false)
```
[the remark this answers is not in the document]
Comment (ynoe) [reply front_end/src/components/cart/ShoppingCart.js]: Resolved in CartReducer.js and ShoppingCart.js

File: front_end/src/components/cart/ShoppingCart.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+  onIncrementCountClick(e) {
+    const productId = e;
+    this.props.incrementItemCount(productId);
+  }
+
+  onDecrementCountClick(e, isRemoveAllItems) {
+    const product = {
+      productId: e,
+      removeAllItems: isRemoveAllItems
```
[the remark this answers is not in the document]
Comment (ynoe) [reply front_end/src/components/cart/ShoppingCart.js]: changed

File: front_end/src/components/cart/ShoppingCart.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    this.onIncrementCountClick = this.onIncrementCountClick.bind(this);
+    this.onDecrementCountClick = this.onDecrementCountClick.bind(this);
+  }
+
+  onIncrementCountClick(e) {
+    const productId = e;
+    this.props.incrementItemCount(productId);
+  }
+
+  onDecrementCountClick(e, isRemoveAllItems) {
```
[the remark this answers is not in the document]
Comment (ynoe) [reply front_end/src/components/cart/ShoppingCart.js]: changed

File: front_end/src/reducers/cartReducer.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+      } else {
+        for (var i = 0; i < state.shoppingCart.length; i++) {
+          if (action.payload.productId == state.shoppingCart[i]._id) {
+            //delete only one occurrence of product being processed
+            state.shoppingCart[i].count--;
+            //If after deletion, count is 0, OR deleting of all occurrences
+            //of product being processed is requested, then remove Product object from cart
+            if (
+              state.shoppingCart[i].count == 0 ||
+              action.payload.removeAllItems
```
[the remark this answers is not in the document]
Comment (ynoe) [reply front_end/src/reducers/cartReducer.js]: changed

File: front_end/public/index.html [hunk 7]
```diff
@@ -12,6 +12,9 @@
   <link rel="manifest" href="%PUBLIC_URL%/manifest.json">
   <link rel="shortcut icon" href="%PUBLIC_URL%/favicon.ico">
 
+  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/latest/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u"
```
Comment (SandroSal) [inline front_end/public/index.html]: what is this?

File: front_end/src/actions/cartActions.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import { ADD_ITEM, REMOVE_ITEM, INCREMENT_ITEM_COUNT } from "./types";
+import axios from "axios";
+
+//do get request at routes/product.js with productName object
+//productName = {name: stringOfProductName}
```
Comment (SandroSal) [inline front_end/src/actions/cartActions.js]: can we delete this comment?

File: front_end/src/actions/queryActions.js [hunk 9]
```diff
@@ -1,6 +1,6 @@
 import { SET_QUERY } from "./types";
 import axios from "axios";
-// do post request at routes/product.js with newQuery object
+// do get request at routes/product.js with newQuery object
 // newQuery = {description : someString }
```
Comment (SandroSal) [inline front_end/src/actions/queryActions.js]: can we delete this comment?

File: routes/api/product.js [hunk 10]
```diff
@@ -6,7 +6,7 @@ const router = express.Router();
 const Product = require("../../models/Product");
 
 // @route GET api/product/search
-// @desc Register user
+// @desc Search for a product using the input string from search bar
```
Comment (SandroSal) [inline routes/api/product.js]: good catch

File: front_end/public/index.html (hunk 7 above)
Comment (minhphan156) [reply front_end/public/index.html]: this is to download style for the shopping cart modal

File: routes/api/product.js (hunk 10 above)
Comment (minhphan156) [reply routes/api/product.js]: thanks for acknowledging

File: front_end/src/actions/cartActions.js (hunk 8 above)
Comment (minhphan156) [reply front_end/src/actions/cartActions.js]: comment deleted

File: front_end/src/actions/queryActions.js (hunk 9 above)
Comment (minhphan156) [reply front_end/src/actions/queryActions.js]: comment deleted

**Pull request merged**